lib/agent0/agent0/hyperdrive/policies/lpandarb.py

```diff
@@ -0,0 +1,642 @@
+tolerance = 1e-18
+MAX_ITER = 10
+MIN_TRADE_AMOUNT = FixedPoint(scaled_value=20)
+
+
+# functions
+def calc_bond_reserves(
+    share_reserves: FixedPoint,
+    share_price: FixedPoint,
+    target_rate: FixedPoint,
+    position_duration: FixedPoint,
+    time_stretch: FixedPoint,
+):
+    """Calculate the amount of bonds that hit the target target rate for the given shares.
+
+    Parameters
+    ----------
+    share_reserves : FixedPoint
+        The amount of share reserves.
+    share_price : FixedPoint
+        The price of the share.
+    target_rate : FixedPoint
+        The target rate.
+    position_duration : FixedPoint
+        The duration of the position.
+    time_stretch : FixedPoint
+        The time stretch factor.
+
+    Returns
+    -------
+    FixedPoint
+        The amount of bonds that hit the target rate.
+    """
+    return share_price * share_reserves * ((fp1 + target_rate * position_duration / fp_seconds_in_year) ** time_stretch)
+
+
+def calc_spot_price_local(
+    initial_share_price: FixedPoint,
+    share_reserves: FixedPoint,
+    share_adjustment: FixedPoint,
+    bond_reserves: FixedPoint,
+    time_stretch: FixedPoint,
+) -> FixedPoint:
+    """Calculate spot price.
+
+    Parameters
+    ----------
+    initial_share_price : FixedPoint
+        The initial price of the share.
+    share_reserves : FixedPoint
+        The amount of share reserves.
+    share_adjustment : FixedPoint
+        The amount of share adjustment.
+    bond_reserves : FixedPoint
+        The amount of bond reserves.
+    time_stretch : FixedPoint
+        The time stretch factor.
+
+    Returns
+    -------
+    FixedPoint
+        The spot price.
+    """
+    effective_share_reserves = share_reserves - share_adjustment
+    return (initial_share_price * effective_share_reserves / bond_reserves) ** time_stretch
+
+
+def calc_apr(
+    share_reserves: FixedPoint,
+    share_adjustment: FixedPoint,
+    bond_reserves: FixedPoint,
+    initial_share_price: FixedPoint,
+    position_duration_seconds: FixedPoint,
+    time_stretch: FixedPoint,
+) -> FixedPoint:
+    """Calculate APR.
+
+    Parameters
+    ----------
+    share_reserves : FixedPoint
+        The amount of share reserves.
+    share_adjustment : FixedPoint
+        The adjustment for share reserves.
+    bond_reserves : FixedPoint
+        The amount of bond reserves.
+    initial_share_price : FixedPoint
+        The initial price of the share.
+    position_duration_seconds : FixedPoint
+        The duration of the position, in seconds.
+    time_stretch : FixedPoint
+        The time stretch factor.
+
+    Returns
+    -------
+    FixedPoint
+        The APR.
+    """
+    annualized_time = position_duration_seconds / fp_seconds_in_year
+    spot_price = calc_spot_price_local(
+        initial_share_price, share_reserves, share_adjustment, bond_reserves, time_stretch
+    )
+    return (fp1 - spot_price) / (spot_price * annualized_time)
+
+
+def calc_k(
```

Comment:
might as well rename this to `calc_k_local` for consistency (to differentiate from what will soon be in the API, but is presently not)